File: gateway/gateway.js
const https = require('https');
const fs = require('fs');
const express = require('express');
const axios = require('axios');
const cors = require('cors');
const { Readable } = require('stream');


const app = express();
const port = process.env.PORT || 3000;

// Read key.js file
const SK = fs.readFileSync('key/key.js', 'utf8');

// CORS middleware to allow requests from any origin
app.use(cors());
// Parse JSON bodies (as sent by API clients)
app.use(express.json());

// Read the certificate and key files
const privateKey = fs.readFileSync('cert/sunsun.dev.key', 'utf8');
const certificate = fs.readFileSync('cert/sunsun.dev.cer', 'utf8');
const ca = fs.readFileSync('cert/sunsun.dev.fullchain.cer', 'utf8');

// Create the credentials object
const credentials = {
    key: privateKey,
    cert: certificate,
    ca: ca
};

// Create the HTTPS server using the credentials and the express app as the request handler function (app)
const httpsServer = https.createServer(credentials, app);

// Route to handle chat requests
// from the frontend app and forward them
// to OpenAI API server using axios library and return the response


// to the frontend app using express response object (res)
app.post('/v1/chat/completions', async (req, res) => {
    try {
        // Extract the data from the request body
        const {
            model,
            messages,
            temperature,
            top_p,
            n,
            stream,
            stop,
            max_tokens,
            presence_penalty,
            frequency_penalty,
            logit_bias,
            user
        } = req.body;

        // Create a Readable stream from the request body
        const streamRequest = new Readable({
            read() {
                this.push(JSON.stringify({
                    model,
                    messages,
                    temperature,
                    top_p,
                    n,
                    stream,
                    stop,
                    max_tokens,
                    presence_penalty,
                    frequency_penalty,
                    logit_bias,
                    user
                }));
                this.push(null);
            }
        });

        // Send a POST request to OpenAI API server
        // and wait for the response
        // and return the response to the frontend app
        const response = await axios.post('https://api.openai.com/v1/chat/completions', streamRequest, {
            headers: {
                'Content-Type': 'application/json',
                'Authorization': SK, // API key
                'Transfer-Encoding': 'chunked',
            },
        });

        // Stream the response back to the frontend app
        // using express response object (res)
        res.setHeader('Content-Type', 'application/json');
        res.setHeader('Transfer-Encoding', 'chunked');
        response.data.pipe(res);
    } catch (error) {
        console.error(error);
        res.status(500).send('Something went wrong');
    }
});


// Start the HTTPS server on the specified port
// httpsServer.listen(port, () => {
//     console.log(`Server listening on port ${port}`);
// });

// Start the HTTP server on the specified port for testing
app.listen(4000, () => {
    console.log(`Server listening on port 4000`);
});

File: gateway/gateway.py
import os
import ssl
import requests
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)

# Allow CORS requests, which are required for the frontend app to be able to
CORS(app)

# Read OPENAI_API_KEY from file
OPENAI_API_KEY = open('./keys/openai_api_key.txt').read()

# Certificate and key files for ssl context
CERT_FILE = './cert/sunsun.dev.cer'
KEY_FILE = './cert/sunsun.dev.key'

# Server port
PORT = 3000

# Create ssl context for HTTPS server using the certificate and key files above
context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
context.load_cert_chain(certfile=CERT_FILE, keyfile=KEY_FILE)


# Route to handle chat requests
# from the frontend app and forward them
# to OpenAI API server using requests library and return the response
# to the frontend app using flask response object (jsonify)
@app.route('/v1/chat/completions', methods=['POST'])
def chat():
    """
    Handle chat requests from the frontend app and forward them to OpenAI API server
    and return the response to the frontend app using flask response object (jsonify)

    :return: Response from OpenAI API server
    """
    try:
        # Get the request body from the client request object (flask)
        request_body = request.json

        # Send the request to OpenAI API server using requests library and get the response object
        response = requests.post('https://api.openai.com/v1/chat/completions', headers={
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + OPENAI_API_KEY
        }, json=request_body)

        # Return the response from OpenAI API to the client using flask response object (jsonify)
        return jsonify(response.json())

    except Exception as e:
        # If there is an error, return an error message to the client
        return jsonify({'error': str(e)})
        

# Start the HTTPS server on the specified port
if __name__ == '__main__':
    app.run(ssl_context=context, port=int(os.environ.get('PORT', PORT)))
